coffee_core/src/coffee.rs

```diff
@@ -288,50 +297,71 @@ impl PluginManager for CoffeeManager {
         self.storage.store(&self.storage_info()).await
     }
 
+    async fn remote_sync(&mut self) -> Result<(), CoffeeError> {
+        for (repo_name, repo) in self.repos.iter_mut() {
+            let repo_path = repo.url().path_string;
+            let repo_url = repo.url().url_string;
+            if !Path::new(&repo_path).exists() {
+                log::warn!("remote with name {repo_name} and URL {repo_url} is not longer present! Coffee is corrupted");
+                return Err(error!(
+                    "coffee storage is out of sync. Please run coffee nurse to resolve the issue"
+                ));
+            }
+        }
+        Ok(())
+    }
+
     async fn add_remote(&mut self, name: &str, url: &str) -> Result<(), CoffeeError> {
+        self.remote_sync().await?;
+        if self.repos.contains_key(name) {
+            return Err(error!("repository with name: {name} already exists"));
+        }
         let url = URL::new(&self.config.root_path, url, name);
         log::debug!("remote adding: {} {}", name, &url.url_string);
         let mut repo = Github::new(name, &url);
         repo.init().await?;
-        self.repos.push(Box::new(repo));
+        self.repos.insert(repo.name(), Box::new(repo));
         log::debug!("remote added: {} {}", name, &url.url_string);
         self.storage.store(&self.storage_info()).await?;
         Ok(())
     }
 
     async fn rm_remote(&mut self, name: &str) -> Result<(), CoffeeError> {
+        self.remote_sync().await?;
         log::debug!("remote removing: {}", name);
-        let Some(index) = self
-            .repos
-            .iter()
-            .position(|x| &*x.to_owned().name() == name) else {
-                return Err(error!("repository with {name} not found"));
-            };
-        let remote_repo = self.repos[index].list().await?;
-        let plugins = self.config.plugins.clone();
-        for plugin in &remote_repo {
-            if let Some(ind) = plugins
-                .iter()
-                .position(|elem| elem.name() == *plugin.name())
-            {
-                let plugin_name = &plugins[ind].name()[..];
-                match self.remove(plugin_name).await {
-                    Ok(_) => {}
-                    Err(err) => return Err(err),
+        match self.repos.get(name) {
+            Some(repo) => {
+                let remote_repo = repo.list().await?;
+                let repo_path = repo.url().path_string;
+                let plugins = self.config.plugins.clone();
+                for plugin in &remote_repo {
+                    if let Some(ind) = plugins
+                        .iter()
+                        .position(|elem| elem.name() == *plugin.name())
+                    {
+                        let plugin_name = &plugins[ind].name()[..];
```

Comment:
```suggestion
                        let plugin_name = &plugins[ind].name().clone()
```

Do not work in this way?

Reply:
Yes this works